File: src/clu/sources/selinux.py
import logging

from clu.provides import Provides
from clu.requires import Requires
from clu.facts import Facts
from clu.sources import Source
from clu.input import text_program
from clu.sources import PARSE_FAIL_MSG

log = logging.getLogger(__name__)


class Selinux(Source):
    def provides(self, provides: Provides) -> None:
        provides["os.selinux.enable"] = self
        provides["os.selinux.mode"] = self

    def requires(self, requires: Requires) -> None:
        requires.programs.extend(["selinuxenabled", "getenforce"])

    def parse(self, facts: Facts) -> None:
        _, rc = text_program("selinuxenabled")
        # man page: "status 0 if SELinux is enabled and 1 if it is not enabled."
        log.debug(f"rc is {rc}")
        if rc == 0:
            facts["os.selinux.enable"] = "True"
        elif rc == 1:
            facts["os.selinux.enable"] = "False"
        else:
            facts["os.selinux.enable"] = PARSE_FAIL_MSG

        data, rc = text_program("getenforce")
        log.debug(f"{data=}")
        facts["os.selinux.mode"] = data.strip() if data else PARSE_FAIL_MSG

File: pkg/facts/sources/selinux.go
// SPDX-FileCopyrightText: 2024 Hunter Matthews
// SPDX-License-Identifier: LGPL-2.1-only

package sources

// Determines SELinux enablement and current mode using `selinuxenabled` and `getenforce`.

import (
	"strings"

	"github.com/huntermatthews/clu/pkg/facts/types"
	"github.com/huntermatthews/clu/pkg/input"
)

// Selinux source collects os.selinux.enable and os.selinux.mode facts.
type Selinux struct{}

var selinuxFacts = map[string]*types.Fact{
	"os.selinux.enable": {Name: "os.selinux.enable", Tier: types.TierOne},
	"os.selinux.mode":   {Name: "os.selinux.mode", Tier: types.TierOne},
}

// Provides registers SELinux fact keys.
func (s *Selinux) Provides(p types.Provides) {
	for name := range selinuxFacts {
		p[name] = s
	}
}

// Requires declares external programs used.
func (s *Selinux) Requires(r *types.Requires) {
	r.Programs = append(r.Programs, "selinuxenabled", "getenforce")
}

// Parse executes the required commands, mapping exit codes and output.
// selinuxenabled: rc 0 -> True, rc 1 -> False, else ParseFailMsg.
// getenforce: trimmed stdout or ParseFailMsg.
func (s *Selinux) Parse(f *types.FactDB) {
	for _, fact := range selinuxFacts {
		fact.Value = types.ParseFailMsg
	}
	defer f.AddAll(selinuxFacts)

	_, rc, _ := input.CommandRunner("selinuxenabled")
	switch rc {
	case 0:
		selinuxFacts["os.selinux.enable"].Value = "True"
	case 1:
		selinuxFacts["os.selinux.enable"].Value = "False"
	}

	if data, _, _ := input.CommandRunner("getenforce"); strings.TrimSpace(data) != "" {
		selinuxFacts["os.selinux.mode"].Value = strings.TrimSpace(data)
	}
}
